### PR TITLE
Warn people that Haml 5 is not fully supported

### LINKED ISSUE #1370: Valid HAML multiline comment produces "Inconsistent indentation" error
### Background

Brakeman version: 4.5.1
Rails version: 5.2.3
Ruby version: 2.6.3

Rails application code: 

```
%p
  -# this comment spans multiple lines.
     Comment lines 2..n are indented with 5 spaces.
     This comment style is valid HAML.
  The comment has ended, we're back to 2 spaces of indentation.
```

See [HAML documentation](http://haml.info/docs/yardoc/file.REFERENCE.html#haml-comments--)

### Issue

```
Error: Inconsistent indentation: 5 spaces used for indentation, but the rest of the document was indented using 2 spaces.
Location: While compiling HAML in /Users/jared/.../redacted.html.haml
```

#### Other Error

Run Brakeman with `--debug` to see the full stack trace.

Stack trace:

```
Error: Inconsistent indentation: 5 spaces used for indentation, but the rest of the document was indented using 2 spaces.
Location: While compiling HAML in /Users/jared/.../redacted.html.haml
/Users/jared/.../redacted.html.haml:999
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:174:in `block in tabs'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:152:in `instance_eval'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:152:in `tabs'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:751:in `block_opened?'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:326:in `haml_comment'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:284:in `silent_script'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:225:in `process_line'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/parser.rb:121:in `parse'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bundle/ruby/2.5.0/gems/haml-4.0.7/lib/haml/engine.rb:66:in `initialize'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/parsers/template_parser.rb:80:in `new'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/parsers/template_parser.rb:80:in `parse_haml'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/parsers/template_parser.rb:27:in `parse_template'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/scanner.rb:87:in `block in parse_files'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/file_parser.rb:29:in `block in read_files'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/file_parser.rb:26:in `each'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/file_parser.rb:26:in `read_files'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/scanner.rb:86:in `parse_files'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/scanner.rb:45:in `process'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman.rb:361:in `scan'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman.rb:80:in `run'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/commandline.rb:133:in `run_brakeman'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/commandline.rb:118:in `regular_report'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/commandline.rb:142:in `run_report'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/commandline.rb:35:in `run'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/lib/brakeman/commandline.rb:20:in `start'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/brakeman-4.5.1/bin/brakeman:10:in `<top (required)>'
/Users/jared/.rbenv/versions/2.6.3/bin/brakeman:23:in `load'
/Users/jared/.rbenv/versions/2.6.3/bin/brakeman:23:in `<top (required)>'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli/exec.rb:74:in `load'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli/exec.rb:28:in `run'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli.rb:463:in `exec'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli.rb:27:in `dispatch'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/cli.rb:18:in `start'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/exe/bundle:30:in `block in <top (required)>'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
/Users/jared/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/bundler-2.0.1/exe/bundle:22:in `<top (required)>'
/Users/jared/.rbenv/versions/2.6.3/bin/bundle:23:in `load'
/Users/jared/.rbenv/versions/2.6.3/bin/bundle:23:in `<main>'
```

### PR DESCRIPTION
**Pull request by jaredbeck**
[Fixes #1370]

I don't have great confidence this is the best implementation, and I'd appreciate feedback.

- How should I write a test for this? Which test file should it go in? 
- In my test, how do I assert that a `Brakeman.notify` occurred?

Thanks.

### CODE REVIEW

**Comment by jaredbeck** [comment]
The fact that the implementation is messy doesn't bother me too much, because we can delete it all when we add support for Haml 5.

File: lib/brakeman/tracker/config.rb [hunk 1]
```diff
@@ -174,6 +191,20 @@ def lower? lhs, rhs
...
     end
 
+    # Brakeman does not support Haml 5 yet. See `check_haml_version`.
+    #
+    # This method is messy because all we have is a version `Requirement` (like
+    # `~> 5.0.3`, or `> 4.99`) The only way I could think of was to loop over
+    # known Haml 5 versions and test whether the `Requirement` is satisfied. I
+    # included '5.99.99' in the list so that this method might stand a chance of
+    # working in the future.
+    def supported_haml_version?
```
Comment (presidentbeef) [inline lib/brakeman/tracker/config.rb]: In many cases, there will be an exact gem version from `Gemfile.lock`.

I am okay with just using `version_between?` which will handle most cases.

File: CHANGES.md [hunk 2]
```diff
@@ -6,6 +6,7 @@
 * Avoid assigning `nil` line numbers to `Sexp`s
 * Add special warning code for custom checks
 * Add call matching by regular expression
+# Warn people that Haml 5 is not fully supported
```
Comment (presidentbeef) [inline CHANGES.md]: Should be `*` not `#`

File: lib/brakeman/tracker/config.rb [hunk 3]
```diff
@@ -111,6 +111,8 @@ def set_rails_version
         @escape_html = true
         Brakeman.notify "[Notice] Escaping HTML by default"
       end
+
+      check_haml_version
```
Comment (presidentbeef) [inline lib/brakeman/tracker/config.rb]: Kind of a weird place for this...but I don't know if there is a less weird place.

File: lib/brakeman/tracker/config.rb (hunk 1 above)
Comment (jaredbeck) [reply lib/brakeman/tracker/config.rb]: `version_between?` expects a single version number, but all we have is a "requirement", right? Consider the requirement `> 4.99`. This requirement will install Haml 5.  It would be incorrect to ignore the operator and only consider the number, 4.99.  `version_between?` would think it is Haml 4.

**Comment by jaredbeck** [comment]
> You can add it to test/tests/brakeman.rb.

Test added, thanks for the guidance!

File: lib/brakeman/tracker/config.rb (hunk 1 above)
Comment (presidentbeef) [reply lib/brakeman/tracker/config.rb]: That is an edge case, yes, but again most applications are going to have a lock file with the exact version, not just the requirement.

File: test/tests/brakeman.rb [hunk 4]
```diff
@@ -2,6 +2,24 @@
...
+    :quiet: false
+    YAML
+    config.close
+    options = {
+      :quiet => :command_line,
+      :config_file => config.path,
+      :app_path => "#{TEST_PATH}/apps/rails5.2",
+      :run_checks => []
+    }
+    _out, err = capture_io { Brakeman.run options }
```
Comment (presidentbeef) [inline test/tests/brakeman.rb]: No need to generate a config file here.
